### PR TITLE
Piping

### PR DESCRIPTION
**Pull request by ericfode**
New Features
- Interfaces have been established for building new sources and outlets
- Tests have been added for the postgres outlet and the redis source and outlet
- Tests are decent examples for how to use this new piping
- Logic for getting locks have been simplified in the redis source
- Lots and Lots more logging for postgres
- How many log lines are dropped is not tracked
- Lots of documentation for go doc!

### CODE REVIEW

File: piping/singleSender_test.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+func TestSingleSender(t *testing.T) {
+	s := NewSingleSender()
+	schan := s.GetSenderChannel()
+	ichan := s.NewOutputChannel("test", 5)
+	schan <- &store.Bucket{
+		Key: store.BKey{Name: "test"}}
+
+	go s.Start()
+	time.Sleep(10)
```
Comment (ericfode) [inline piping/singleSender_test.go]: Get rid of the sleep

File: postgres_outlet/main.go [hunk 2]
```diff
@@ -17,6 +15,7 @@ var (
 	workers         int
 	processInterval int
 	lockTTL         uint64
+	database_url    string
```
Comment (ericfode) [inline postgres_outlet/main.go]: Delete this line

File: store/redis_test.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package store
```
Comment (ericfode) [inline store/redis_test.go]: Delete this file

**Comment by ryansmith3136** [comment]
@ericfode I don't understand your inline comments. (e.g. Delete this file, Delete this line, etc...) Can you elaborate?

**Comment by ericfode** [comment]
Todos

**Comment by ryansmith3136** [comment]
I see. You should consider using check lists in a gist comment to better communicate these todos. Should I hold off consideration of this PR until you finish the TODOs?

**Comment by ericfode** [comment]
Please do look at it and leave any changes that will need to be made for it
to be accepted and I will take care of it all at once. All of the things
that are commented on by me in there are cosmetic or left overs from
debugging.

**Comment by ryansmith3136** [comment]
@ericfode Your code here has inspired me. I have borrowed a lot of the concepts you laid out and merged them into master. While this particular set of changes can't be directly integrated, your ideas live on. Thanks for the hard work!

**Pull request closed without merging**